### PR TITLE
CA-1229 + CA-1339: Allow users to update the billing account for billing projects

### PR DESCRIPTION
**Pull request by MatthewBemis**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceV2.scala [hunk 1]
```diff
@@ -53,6 +53,22 @@ trait BillingApiServiceV2 extends UserInfoDirectives {
...
+            put {
+              entity(as[UpdateRawlsBillingAccountRequest]) { updateProjectRequest =>
+                complete{
+                  userServiceConstructor(userInfo).updateBillingProjectBillingAccount(RawlsBillingProjectName(projectId), updateProjectRequest).map(_ => StatusCodes.OK)
+                }
+              }
+            } ~
+              delete {
+                complete {
+                  userServiceConstructor(userInfo).deleteBillingAccount(RawlsBillingProjectName(projectId)).map(_ => StatusCodes.OK)
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceV2.scala:67]: I originally had these as `No Content` but it looks like `v2_workspaces` returns `OK` with a body.

Any objections to returning `OK` but with no body for now? I figure that is probably the path of least resistance for harmonizing these two epics once they're both live.

File: core/src/main/resources/swagger/api-docs.yaml [hunk 2]
```diff
@@ -120,6 +120,76 @@ paths:
             - profile
             - https://www.googleapis.com/auth/cloud-billing
       x-codegen-request-body-name: createProjectRequest
+  /api/billing/v2/{projectId}/billingAccount:
```
Comment (MatthewBemis) [inline core/src/main/resources/swagger/api-docs.yaml:123]: This swagger change is ripped right from `v2_workspaces`. Only change is that I return `OK` with no body instead of `OK` with a `RawlsBillingProjectResponse`. I left another comment on that in the route layer

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 3]
```diff
@@ -461,6 +461,36 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     }
   }
 
+  def updateBillingProjectBillingAccount(billingProjectName: RawlsBillingProjectName, updateAccountRequest: UpdateRawlsBillingAccountRequest): Future[Unit] = {
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:464]: `v2_workspaces` returns a `PerRequestMessage` but I'd like to stop proliferating that pattern so I don't use that here. If y'all would find it easier to mesh with, I am happy to make this return a `PerRequestMessage` as well and then take care of it all in my currently-paused [PerRequest refactor](https://github.com/broadinstitute/rawls/pull/1408) after PPW goes live.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 4]
```diff
@@ -461,6 +461,36 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
...
+    requireProjectAction(billingProjectName, SamBillingProjectActions.updateBillingAccount) {
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:467]: From what I can tell, this permission check is not present in the v2_workspaces branch. Should I just file a Jira for it to be added?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 3 above)
Comment (s-rubenstein) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:464]: Out of curiosity, what's the rationale in removing that pattern?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceV2.scala (hunk 1 above)
Comment (s-rubenstein) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceV2.scala:67]: That seems fine to me, for the time being. I generally prefer returning the updated object to returning an empty body, because it gives clients the ability to base their updates on the resulting object, rather than guessing or making a fresh api call.

**Review by s-rubenstein — APPROVED** [review]
This looks reasonable to me. Mostly just philosophical conversations about response types, which definitely aren't blocking.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 3 above)
Comment (MatthewBemis) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:464]: The main reason is that it obfuscates the return types of many of the methods in the service layer, which can make the code harder to reason about and test. It also causes status codes to bleed more deeply into the business logic when preferably that would be taken care of in the API route layer.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceV2.scala (hunk 1 above)
Comment (MatthewBemis) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceV2.scala:67]: Sorry forgot to respond to this. The reason that I first chose to return nothing here is that the format of what I would be able to return here would be quite different than what the endpoint will return once v2_workspaces is merged. I could get close, but it would require bringing in a bunch of other code from v2_workspaces.

My thinking is that it is better to go from no response body to _some_ response body, rather than going from some response body that sets expectations for field names, values, etc and then switching to something else.

In general, I totally agree with you about trying to return the updated object when applicable.

Comment (gpolumbo-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceV2.scala:67]: I think I'm in agreement with you Matt.  Better to return nothing for now and then add to it.  Technically, we might consider that an API version bump, but I think it's ok here.  If we actually change the contents of what is returned, then I think we'd need to version the API again with PPW, and I don't think anybody wants to do that.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala [hunk 5]
```diff
@@ -401,4 +404,132 @@ class UserServiceSpec extends AnyFlatSpecLike with TestDriverComponent with Mock
...
+      Await.result(userService.updateBillingProjectBillingAccount(billingProject.projectName, newBillingAccountRequest), Duration.Inf) shouldEqual ()
+
+      val spendReportDatasetInDb = runAndWait(dataSource.dataAccess.rawlsBillingProjectQuery.filter(_.projectName === billingProject.projectName.value).map(row => (row.spendReportDataset, row.spendReportTable)).result)
+
+      //assert that the spend report configuration has been fully cleared
+      spendReportDatasetInDb.head shouldEqual (None, None)
+
+      verify(mockGcsDAO, times(1)).setGoogleProjectBillingAccount(ArgumentMatchers.eq(googleProjectName), any[Option[RawlsBillingAccountName]], any[UserInfo])(any[ExecutionContext])
+    }
+  }
```
Comment (gpolumbo-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:432]: Matt, I know you pulled all this over from the PPW branch, but I am inclined to add a test for when we call the Update Billing Account endpoint with an empty/blank Billing Account field.  At the moment, I think that's something somebody could do and I think it would result in a failure when checking if we have access to the Billing Account, and I guess that's fine, but that's the sort of thing we should probably test because callers of the API may come to depend on that behavior.

Comment (MatthewBemis) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:432]: These tests were actually not lifted from the PPW branch. I can look at adding that test case, though.

Comment (MatthewBemis) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:432]: I am writing a test for this scenario and truthfully I don't think that it's valuable. This would test Google's behavior, which I can certainly mock out, but the assertion would just be that the mock did what I told it to do.

Comment (gpolumbo-broad) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:432]: That makes sense.  I guess from my perspective, the long story short is that I'd rather see us have some kind of validation check(s) on the Billing Account that was submitted.  And whether we do that is based on standards/expectations we already set for other fields maybe as well as how the code behaves today.  

What validations should we check?  Do we know the min/max lengths and maybe regex that a Billing Account should match?  We certainly know that if it's an empty string, it's not valid, and I'd rather explicitly catch that.  

My preference would be that overall, we do good validation on values entered rather than counting on making the call to Google to catch whether we entered a bad value.  But I'm ok leaving it up to you to decide if you want to add this at this time.  It wasn't specified in the original ticket, and we can add it later if people are concerned about the behavior or the UX proves problematic.  In general, I do just want to put in my plug for doing good validation when we can.

tl;dr - my preference is to do good validation, but I'm OK with you making the call on this one

**Review by gpolumbo-broad — APPROVED** [review]
